### ISSUE TITLE
Firewall exceptions don't apply when running through dotnet command

### ISSUE THREAD

**Issue by Telavian** [dotnet/core#2609]
I originally filled this issue at StackOverflow [here](https://stackoverflow.com/questions/55766545/is-it-possible-to-add-firewall-exceptions-to-a-net-core-application) however I think it makes more sense to be here also.

I have some .net core microservices that I am deploying to remote machines through [nomad](https://www.nomadproject.io/). At startup each service gets a dynamic port and uses that.

The steps are essentially:
1. Build project as publish
2. Zip output
3. Create nomad job script
4. Deploy through nomad.
5. Nomad runs execution script.

This whole process works and I like it. The issue is though that each service needs to communicate with each other and is using dynamic ports so a firewall exception is needed.
I tried adding a firewall exception for the dll however that doesn't seem to work.
However if I compile to an exe and add the exception then it does.

I am using [Fabio](https://fabiolb.net/) as a reverse proxy however that takes the request and forwards it directly to the machine:port where the service actually is.

If using dotnet xyz.dll to execute .net core applications is the suggested way then how to also support dynamic firewall rules?

If it is helpful this is my execution script that nomad uses:
```
netsh advfirewall firewall add rule name="Allow %1" dir=in action=allow program="%2"
dotnet %2
netsh advfirewall firewall delete rule name="Allow %1" dir=in
```

In summary above script adds the firewall rules which don't work if %2 is a dll.
If %2 is an exe then firewall works (dotnet %2 becomes just %2 since it is exe)

**Comment by karelz** [comment]
Is it limitation of the firewall that is not able to find process based on DLL name? (that's what the program is I assume)
Maybe I am missing something ...
What would you suggest that dotnet should change?

Can you just hack it and rename DLL to EXE - will that make it work?

**Comment by Telavian** [comment]
That is a good idea to try the rename. I honestly didn't think of that however I doubt it would work.

I think in a developers mind dotnet xyz.dll runs the program xyz which should obey all "program" rules like firewall exceptions. Since xyz is the program in this case.

However it seems that in reality dotnet.exe is run takes the IL in xyz.dll and executes that. This all makes sense however what to do with firewall rules?

The network requests are coming from dotnet not xyz.dll so the firewall ignores those rules.
I would expect dotnet to interact with the firewall in an intelligent manner so it knows this information and can react to it.

I can try the rename. If it works I will be highly surprised.

**Comment by karelz** [comment]
Then register the firewall for dotnet.exe as it is the main EXE.
I don't think .NET should know anything about firewalls, etc.

**Comment by Telavian** [comment]
Registering dotnet.exe in the firewall is a massive security hole.
I do agree that dotnet.exe shouldn't know about the firewall however maybe the firewall should know about dotnet.

Seems like there is not clear cut answer to this.
However suggesting .net core be run via dotnet.exe and firewall exceptions be limited to the application instead of port means network access is highly limited.

**Comment by karelz** [comment]
How does Java work in this case with this firewall? I bet they ran into similar problems ... or any other runtime using runner.exe.

**Comment by Telavian** [comment]
Good question. I can investigate.

**Comment by Telavian** [comment]
It appears they have the same issue which makes sense.
Firewall should allow finer grain permissions especially when you have an app runner however that would probably be a mess.

Limiting it to exe makes it clear and easy however it limits the portability of dynamic network enabled .net core application.

Anyways thanks for trying to understand this.
